scipy/integrate/tests/test_quadrature.py

```diff
@@ -457,3 +459,398 @@ def func(x):
+
+        message = '...must be reals.'
+        with pytest.raises(ValueError, match=message):
+            _tanhsinh(f, 1+1j, f.b)
+        with pytest.raises(ValueError, match=message):
+            _tanhsinh(f, 0, np.nan)
+        with pytest.raises(ValueError, match=message):
+            _tanhsinh(f, 0, f.b, atol='ekki')
+        with pytest.raises(ValueError, match=message):
+            _tanhsinh(f, 0, f.b, rtol=pytest)
+
+        message = '...must be non-negative and finite.'
+        with pytest.raises(ValueError, match=message):
+            _tanhsinh(f, 0, f.b, rtol=-1)
+        with pytest.raises(ValueError, match=message):
+            _tanhsinh(f, 0, f.b, atol=np.inf)
+
+        message = '...may not be positive infinity.'
+        with pytest.raises(ValueError, match=message):
+            _tanhsinh(f, 0, f.b, rtol=np.inf, log=True)
+        with pytest.raises(ValueError, match=message):
+            _tanhsinh(f, 0, f.b, atol=np.inf, log=True)
+
+        message = '...must be integers.'
+        with pytest.raises(ValueError, match=message):
+            _tanhsinh(f, 0, f.b, maxlevel=object())
+        with pytest.raises(ValueError, match=message):
+            _tanhsinh(f, 0, f.b, maxfun=1+1j)
+        with pytest.raises(ValueError, match=message):
+            _tanhsinh(f, 0, f.b, minlevel="migratory coconut")
+
+        message = '...must be non-negative.'
+        with pytest.raises(ValueError, match=message):
+            _tanhsinh(f, 0, f.b, maxlevel=-1)
+        with pytest.raises(ValueError, match=message):
+            _tanhsinh(f, 0, f.b, maxfun=-1)
+        with pytest.raises(ValueError, match=message):
+            _tanhsinh(f, 0, f.b, minlevel=-1)
+
+    @pytest.mark.parametrize("limits, val", [
+        [(0, np.inf), 0.5],  # b infinite
+        [(-np.inf, 0), 0.5],  # a infinite
+        [(-np.inf, np.inf), 1],  # a and b infinite
+        [(np.inf, -np.inf), -1],  # flipped limits
+        [(1, -1), stats.norm.cdf(-1) -  stats.norm.cdf(1)],  # flipped limits
+    ])
+    def test_integral_transforms(self, limits, val):
+        # Check that the integral transforms are behaving for both log and
+        # normal integration
+        dist = stats.norm()
+
+        res = _tanhsinh(dist.pdf, *limits)
+        assert_allclose(res.integral, val)
+
+        logres = _tanhsinh(dist.logpdf, *limits, log=True)
+        assert_allclose(np.exp(logres.integral), val)
+        # Transformation should not make the result complex unnecessarily
+        assert (np.isreal(logres.integral) if val > 0
+                else np.iscomplex(logres.integral))
+
+        assert_allclose(np.exp(logres.error), res.error, atol=1e-16)
+
+    # 15 skipped intentionally; it's very difficult numerically
+    @pytest.mark.parametrize('f_number', range(1, 15))
+    def test_basic(self, f_number):
+        f = getattr(self, f"f{f_number}")
+        res = _tanhsinh(f, 0, f.b)
+        assert_allclose(res.integral, f.ref)
```

Comment:
For easy problems, check status, and confirm that estimated error is less than required. (Maybe less than actual error?)